### ISSUE TITLE
No build script to automatically link for embedded dart vm

### ISSUE THREAD

**Issue by jc324** [DoumanAsh/dart-sys#1]
Hi,

I keep getting this link error:
```
= note: vm.histy2mu0occs4k.rcgu.o : error LNK2019: unresolved external symbol Dart_VersionString referenced in function _ZN2vm4main17h0ad19c8a20b6b3a7E
...
fatal error LNK1120: 1 unresolved externals
```

Are these bindings working on your side?

**Comment by DoumanAsh** [comment]
How do you build?
Please tell me version of library and describe build process?

`Dart_VersionString ` should definitely be present in any version hence I can assume that you do not link dart runtime library

P.s. I use this library in my work project with flutter but I build shared library hence it is linked only when flutter application is executed

**Comment by jc324** [comment]
I'm not executing it in a flutter application, but rather looking to get it working in a normal rust project. Here are the files. Thanks.

Cargo.toml
```toml
[package]
name = "demo"
version = "0.1.0"
edition = "2018"
[dependencies]
dart-sdk-sys = "2.14.1"
```

main.rs
```rust
extern crate dart_sdk_sys;

pub use dart_sdk_sys::Dart_VersionString;

fn main() {
    let version_c = unsafe { Dart_VersionString() };

    println!("DART VERSION: {:?}", version_c);
}
```

Error:
```
cargo run
   Compiling libc v0.2.102
   Compiling dart-sdk-sys v2.14.1
   Compiling demo v0.1.0 (C:\Users\jc324\Documents\Projects\demo)
error: linking with `link.exe` failed: exit code: 1120
  |
  = note: "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.28.29333\\bin\\HostX64\\x64\\link.exe" "/NOLOGO" "/NXCOMPAT" "/LIBPATH:C:\\Users\\jc324\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "C:\\Users\\jc324\\Documents\\Projects\\demo\\target\\debug\\deps\\demo.1bmpd9ksufgte5ew.rcgu.o" "C:\\Users\\jc324\\Documents\\Projects\\demo\\target\\debug\\deps\\demo.1e18l8pk67ks1wnx.rcgu.o" "C:\\Users\\jc324\\Documents\\Projects\\demo\\target\\debug\\deps\\demo.1hst314g5ktcrvui.rcgu.o" "C:\\Users\\jc324\\Documents\\Projects\\demo\\target\\debug\\deps\\demo.474b5u4opml8nrt2.rcgu.o" "C:\\Users\\jc324\\Documents\\Projects\\demo\\target\\debug\\deps\\demo.4k34r1gjcc5x2ztz.rcgu.o" "C:\\Users\\jc324\\Documents\\Projects\\demo\\target\\debug\\deps\\demo.4udpt5v1tsuz66mr.rcgu.o" "C:\\Users\\jc324\\Documents\\Projects\\demo\\target\\debug\\deps\\demo.54gnxsl1bl0ntkg.rcgu.o" "C:\\Users\\jc324\\Documents\\Projects\\demo\\target\\debug\\deps\\demo.5fkv5m9b4i2gpa5u.rcgu.o" "C:\\Users\\jc324\\Documents\\Projects\\demo\\target\\debug\\deps\\demo.sm4oy2zbvyck46a.rcgu.o" "/OUT:C:\\Users\\jc324\\Documents\\Projects\\demo\\target\\debug\\deps\\demo.exe" "C:\\Users\\jc324\\Documents\\Projects\\demo\\target\\debug\\deps\\demo.2vkoyg20ywrbuez6.rcgu.o" "/OPT:REF,NOICF" "/DEBUG" "/NATVIS:C:\\Users\\jc324\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\intrinsic.natvis" "/NATVIS:C:\\Users\\jc324\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\liballoc.natvis" "/NATVIS:C:\\Users\\jc324\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\libcore.natvis" "/NATVIS:C:\\Users\\jc324\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\libstd.natvis" "/LIBPATH:C:\\Users\\jc324\\Documents\\Projects\\demo\\target\\debug\\deps" "/LIBPATH:C:\\Users\\jc324\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "C:\\Users\\jc324\\Documents\\Projects\\demo\\target\\debug\\deps\\libdart_sdk_sys-9f32d61ccfc0d841.rlib" "C:\\Users\\jc324\\Documents\\Projects\\demo\\target\\debug\\deps\\liblibc-a02f6a92f87fd9bf.rlib" "C:\\Users\\jc324\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libstd-0087fff1d6c84113.rlib" "C:\\Users\\jc324\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libpanic_unwind-e9a9adcb7b282f86.rlib" "C:\\Users\\jc324\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\librustc_demangle-c17bc77d789f8359.rlib" "C:\\Users\\jc324\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libhashbrown-1f739c53c46039b3.rlib" "C:\\Users\\jc324\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\librustc_std_workspace_alloc-b12f9afb50af9e82.rlib" "C:\\Users\\jc324\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libunwind-4813d8391ae0f1e1.rlib" "C:\\Users\\jc324\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcfg_if-a6dfd598b03abbd4.rlib" "C:\\Users\\jc324\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\liblibc-8e04de9ca6363561.rlib" "C:\\Users\\jc324\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\liballoc-05126e82a7fe403c.rlib" "C:\\Users\\jc324\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\librustc_std_workspace_core-7f3e1ef8417012fa.rlib" "C:\\Users\\jc324\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcore-df80c014103f27c5.rlib" "C:\\Users\\jc324\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcompiler_builtins-f404d34c325d6d79.rlib" "advapi32.lib" "ws2_32.lib" "userenv.lib" "msvcrt.lib"
  = note: demo.54gnxsl1bl0ntkg.rcgu.o : error LNK2019: unresolved external symbol Dart_VersionString referenced in function _ZN4demo4main17ha104b92318a8665bE
          C:\Users\jc324\Documents\Projects\demo\target\debug\deps\demo.exe : fatal error LNK1120: 1 unresolved externals


error: aborting due to previous error

error: could not compile `demo`

To learn more, run the command again with --verbose.
```

**Comment by DoumanAsh** [comment]
Ok, then you need to link to dart library.
In case of flutter it is `libflutter.so`
but if you want to do pure dart then we need to link it with [dart sdk](https://dart.dev/get-dart)

I didn't make build script for that specifically because in case of flutter you don't actually need to do it manually.

So please describe your use case:
- Do you want to build library that should be used by dart?
- Do you want to build executable that uses embedded dart vm?

If it is first use case, you should build your library as `cdylib`. But I can make build script for this bindings where you can provide Dart SDK path.
If it is second case then I need to hear more details of what API you require.
My current bindings do not include embedded VM API.

TLDR:
You need to write build.rs with following instructions:
```
    let dart_path = //Path pointing to Dart SDK
    println!(r#"cargo:rustc-link-search=native={}"#, dart_path.join("bin").to_str().unwrap());
    println!(r"cargo:rustc-link-lib=dart");
```

**Comment by DoumanAsh** [comment]
If you can, ping me on discord `Douman#7815` for faster communication

**Comment by DoumanAsh** [comment]
@jc324 
I'll write down steps on how you're _supposed_ to use dart FFI:
1. Build C shared library (in case of rust you need to make your library as `cdylib`)
2. Load your library using dart ffi `ffi.DynamicLibrary.open`
3. If successful you can use `lookup` function to get function pointer.

**Comment by DoumanAsh** [comment]
I've added detailed instructions for use cases of Dart FFI 
https://github.com/DoumanAsh/dart-sys#use-cases

Please let me know if you need further help after reading onto it.
Note that generally you're not supposed to use dart SDK bindings outside of flutter/dart application

**Comment by jc324** [comment]
I already ended up writing custom build script based on [dart-sys](https://crates.io/crates/dart-sys). I'm not trying to create a dart native extension, instead I'm interested in the second use case, building an exe that embeds dart vm. I'll join your discord. 👍

**Comment by DoumanAsh** [comment]
I'm closing this issue as I'm only interested in using dart sdk to to support native extensions.
As long as embedded VM API is not stable, I'll refrain from trying to include it
